### DISCUSSION TITLE
portable-pty: how to know whether it's stdout or stderr?

### DISCUSSION THREAD

**Discussion by jjangga0214** [wezterm/wezterm#3948] [Q&A]
Hi!

I am using portable-pty.
Let's say I am using this basic snippet.

```rs
 // Other context is omitted for brevity. 
 // Name of the variable binding `pair` are identical to the official example code.
 let reader = pair.master.try_clone_reader().unwrap();
 std::thread::spawn(move || {
    let reader = BufReader::new(reader);
    for line in reader.lines() {
      match line {
        Ok(value) => {
          // How can I let `do_something` know if the `value` is stdout or stderr?
          do_something(value); 
        }
        _ => break,
      }
    }
  });
```

I am reading output line by line.
But how can I know whether the line is stdout or stderr?
For example, an arbitrary command may produce stdout for a while, but suddenly outputs stderr in the middle of execution.
How can I know it?

**Comment by wez** [comment]
It's not possible; TTYs and PTYs don't have a concept of stdout or stderr, just a single output stream.

Programs spawned into a PTY (usually, they are shell programs) are spawned with:

* `stdin` is a descriptor that maps to the input of the PTY
* `stdout` is a descriptor that maps to the output of the PTY
* `stderr` is a duplicate of the descriptor that maps to the output of the PTY

It is then usually the shell that allows the user to choose to redirect `stdout` or `stderr` to an alternative stream.
